internal/cmd/commands/connect/credentials.go

```diff
@@ -0,0 +1,62 @@
+package connect
+
+import (
+	"errors"
+
+	"github.com/hashicorp/boundary/api/targets"
+	"github.com/hashicorp/boundary/internal/credential"
+	"github.com/mitchellh/mapstructure"
+)
+
+type usernamePasswordCredentials struct {
+	Username string `mapstructure:"username"`
+	Password string `mapstructure:"password"`
+}
+
+// TODO: this should support multiple types of credentials and return an interface instead
+// of only usernamepassword. Each subcommand can use the returned credentials as it needs.
+func parseCredentials(creds []*targets.SessionCredential) ([]usernamePasswordCredentials, error) {
+	if creds == nil {
+		return nil, nil
+	}
+	var out []usernamePasswordCredentials
+	for _, cred := range creds {
+		if cred.CredentialSource == nil {
+			return nil, errors.New("missing credential source")
+		}
+
+		if cred.CredentialSource.CredentialType == string(credential.UsernamePasswordType) {
+			username, ok := cred.Credential["username"].(string)
```

Comment:
Given you have `mapstructure` tags, why not just `mapstructure.Decode` into the struct? Then you don't need to do the validity checking on the map.